wolfcrypt/src/port/Espressif/esp32_sha.c

```diff
@@ -0,0 +1,437 @@
+    esp_wait_until_idle();
+    /* load message data into hw */
+    for(i=0;i<((len)/(sizeof(word32)));++i){
+        DPORT_REG_WRITE(SHA_TEXT_BASE+(i*4),*(data+i));
+    }
+    /* notify hw to start process */
+    esp_sha_start_process(ctx, address);
+
+    ESP_LOGV(TAG, "leave esp_process_block");
+}
+/*
+* retreive sha digest from memory
+*/
+void esp_digest_state(WC_ESP32SHA* ctx, byte* hash, enum SHA_TYPE sha_type)
+{
+    /* registers */
+    word32 SHA_LOAD_REG = SHA_1_LOAD_REG;
+    word32 SHA_BUSY_REG = SHA_1_BUSY_REG;
+
+    ESP_LOGV(TAG, "enter esp_digest_state");
+
+    /* sanity check */
+    if(sha_type == SHA_INVALID) {
+        ESP_LOGE(TAG, "unexpected error. sha_type is invalid.");
+        return;
+    }
+    
+    SHA_LOAD_REG += (sha_type << 4);
+    SHA_BUSY_REG += (sha_type << 4);
+
+    if(ctx->isfirstblock == 1){
+        /* no hardware use yet. Nothing to do yet */
+        return ;
+    }
+
+    {
+        /* wait until idle */
+        esp_wait_until_idle();
+
+        /* LOAD final digest */
+        DPORT_REG_WRITE(SHA_LOAD_REG, 1);
+        /* wait until done */
+        while(DPORT_REG_READ(SHA_BUSY_REG) == 1){ }
+
+        esp_dport_access_read_buffer((word32*)(hash), SHA_TEXT_BASE,
+                                 esp_sha_digest_size(sha_type)/sizeof(word32));
+
+#if defined(WOLFSSL_SHA512) || defined(WOLFSSL_SHA384)
+        if(sha_type==SHA2_384||sha_type==SHA2_512) {
+            word32  i;
+            word32* pwrd1 = (word32*)(hash);
+            /* swap value */
+            for(i = 0; i <WC_SHA512_DIGEST_SIZE/4; i+=2 ) {
+                pwrd1[i]  ^= pwrd1[i+1];
+                pwrd1[i+1]^= pwrd1[i];
+                pwrd1[i]  ^= pwrd1[i+1];
+            }
+        }
+#endif
+    }
+
+    ESP_LOGV(TAG, "leave esp_digest_state");
+}
+
+#ifndef NO_SHA
+/*
+* sha1 process
+*/
+int esp_sha_process(struct wc_Sha* sha)
+{
+    int ret = 0;
+
+    ESP_LOGV(TAG, "enter esp_sha_process");
+    
+    word32 SHA_START_REG = SHA_1_START_REG;
+    
+    esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                        WC_SHA_BLOCK_SIZE);
+
+    ESP_LOGV(TAG, "leave esp_sha_process");
+    return ret;
+}
+/*
+* retrieve sha1 digest
+*/
+int esp_sha_digest_process(struct wc_Sha* sha, byte blockproc)
+{
+    int ret = 0;
+    
+    ESP_LOGV(TAG, "enter esp_sha_digest_process");
+
+    if(blockproc) {
+        word32 SHA_START_REG = SHA_1_START_REG;
+
+        esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                            WC_SHA_BLOCK_SIZE);
+    }
+
+    esp_digest_state(&sha->ctx, (byte*)sha->digest, SHA1);
+
+    ESP_LOGV(TAG, "leave esp_sha_digest_process");
+
+    return ret;
+}
+#endif /* NO_SHA */
+
+
+#ifndef NO_WOLFSSL_SHA256
+/*
+* sha256 process
+*/
+int esp_sha256_process(struct wc_Sha256* sha)
+{
+    int ret = 0;
+    
+    ESP_LOGV(TAG, "enter esp_sha256_process");
+    
+    word32 SHA_START_REG = SHA_1_START_REG + 16;
```

Comment:
Please move variable declaration to top of function. Is it possible to replace `16` with macro? Is that supposed to be block size or is it a hardware register offset?

Reply:
Move the declaration and use macro instead of value.